Taking ownership of memmove.S

j1ah0ng opened the pull request:
~~**This pull is incomplete. Do not merge.**~~ Pull is ready for merge, pending review.

I'm beginning development on `memmove.S`.

j1ah0ng committed:
Taking ownership of memmove.S

j1ah0ng committed:
Initial implementation of memmove

sushisharkjl commented:
Ready to merge, all test pass!

j1ah0ng committed:
Fixing an issue with `memmove` on overlapping pointers and updating unit tests to check this case

Achierius commented on src/asm/memmove.S:
```diff
@@ -8,5 +8,73 @@ SECTION .TEXT
+
+qwords_in_rev:
+    sub rdi, 3          ; rdi -= 3
+    sub rsi, 3          ; rsi -= 3
```
Not gonna lie I don't get why this offset of 3 is here -- it's not scaling ofc, but I don't get why a constant offset would matter when you're calling a rep-prefixed instr anyways

sushisharkjl replied:
When we call `movsq <DEST>, <SRC>` the QWORD is written to the 4-byte block _beginning_ at `<DEST>`--since we've already added a 1B offset in the reverse case, we need to add 3 more bytes of offset so we don't segfault over the end of our block.

sushisharkjl replied:
This has just reminded me that QWORDS is eight bytes. Fix incoming

Achierius commented on src/asm/memmove.S:
```diff
@@ -8,5 +8,73 @@ SECTION .TEXT
+    sub rsi, 1          ; rsi += rcx - 1
+
+not_above:
+    cld                 ; Unset direction flag
```
til very cool

Achierius reviewed (changes requested):
The only necessary comment is the first one (re: the addition to ASMFLAGS), the rest are just thoughts on the implementation / style / etc.

sushisharkjl commented on Makefile:
```diff
@@ -8,7 +8,7 @@ FORCE:
 
 ASMR = nasm
-ASMFLAGS = -X gnu -f elf64
+ASMFLAGS = -X gnu -f elf64 -g
```
I was using GDB earlier---it'll be removed when I finish up ofc ofc

sushisharkjl commented on src/asm/memmove.S:
```diff
@@ -8,5 +8,73 @@ SECTION .TEXT
+not_above:
+    cld                 ; Unset direction flag
+
+qwords_check:
```
Nice! I'll go through and use that---looks a lot cleaner.

sushisharkjl commented on src/asm/memmove.S:
```diff
@@ -8,5 +8,73 @@ SECTION .TEXT
+    ; - rdx:    Raw n. bytes to copy
+
+    push rdi            ; Save the dest addr
+    push rbx            ; We need another register
```
I was trying to avoid the 64-bit registers (but since I'm going to try using streaming extensions, this seems moot at least for that implementation)

Achierius replied:
rdi and rbx are also 64 bit, are they not? edi and ebx would be their 32-bit equivalents
STREAMING EXTENSIONS HYPE

j1ah0ng committed:
Fixing QWORD size mismatch

j1ah0ng committed:
Minimum viable implementation.

This is an implementation of the naive version of `memmove`. We use simple forwards/backwards switching.

j1ah0ng commented:
@Achierius OK to merge!

j1ah0ng committed:
README updates

j1ah0ng committed:
Removing debugging changes/files and reverting Makefile.

Achierius committed:
Merge branch 'master' into dev_memmove
